ekncontent/ekncontent/eknc-domain.c

```diff
@@ -850,6 +851,55 @@ eknc_domain_query_finish (EkncDomain *self,
 }
 
 /**
+ * eknc_domain_read_uri:
+ * @self: domain
+ * @uri: the ekn uri to read
+ * @bytes: (out) (transfer none) (allow-none): return location for the contents GBytes
+ * @mime_type: (out) (transfer full) (allow-none): return location for the content mime type
+ * @error: #GError for error reporting
+ *
+ * Reads the contents of a ekn uri and returns a GBytes of the contents and the
+ * contents mime type, if the ekn uri contents was found.
+ *
+ * Returns: true if the uri was successfully searched for, false if an error occurred
+ */
+gboolean
+eknc_domain_read_uri (EkncDomain *self,
+                      const gchar *uri,
+                      GBytes **bytes,
+                      const gchar **mime_type,
+                      GError **error)
+{
+  g_return_val_if_fail(uri && g_str_has_prefix (uri, "ekn://"), FALSE);
+  g_auto(GStrv) tokens = g_strsplit (uri + strlen("ekn://"), "/", -1);
+  g_return_val_if_fail(tokens && tokens[0] && tokens[1], FALSE);
+
+  EosShardRecord *record = NULL;
+  /* iterate over all shards until we find a matching record */
+  for (GSList *l = self->shards; l && !record; l = g_slist_next (l))
+    record = eos_shard_shard_file_find_record_by_hex_name (l->data, tokens[1]);
+
+  if (!record)
+    return TRUE;
+
+  EosShardBlob *blob = record->data;
+  /* Use resource, if present */
+  if (tokens[2])
+    blob = eos_shard_record_lookup_blob (record, tokens[2]);
+
+  if (!blob)
+    return TRUE;
+
+  if (mime_type)
+    *mime_type = eos_shard_blob_get_content_type (blob);
```

Comment:
I think you have to `g_strdup()` this and remove `const` from `mime_type` in the function signature, it's transfer none so otherwise it won't outlive the `blob` or the `record`, and you have it annotated as transfer full.